src/TpetraSegregatedLinearSystem.C

```diff
@@ -0,0 +1,1829 @@
+
+    if (localId > maxSharedNotOwnedRowId_) {
+      throw std::runtime_error("logic error: localId > maxSharedNotOwnedRowId");
+    }
+
+    const double dval = useOwned ? diag_value : 0.0;
+    // Adjust the LHS; zero out all entries (including diagonal)
+    matrix->getLocalRowView(actualLocalId, indices, values);
+    const size_t rowLength = values.size();
+    if (rowLength > 0) {
+      new_values.resize(rowLength);
+      for (size_t i=0; i < rowLength; i++) {
+        new_values[i] = (indices[i] == localId) ? dval : 0.0;
+      }
+      local_matrix.replaceValues(actualLocalId, &indices[0], rowLength, new_values.data(), internalMatrixIsSorted);
+    }
+
+    // Replace RHS residual entry = 0.0
+    Teuchos::RCP<LinSys::MultiVector> rhs = useOwned ? ownedRhs_ : sharedNotOwnedRhs_;
+    for (unsigned d = beginPos; d < endPos; ++d) {
+      rhs->replaceLocalValue(actualLocalId, d, rhs_residual);
+    }
+  }
+}
+
+void TpetraSegregatedLinearSystem::loadComplete()
+{
+  // LHS
+  Teuchos::RCP<Teuchos::ParameterList> params = Teuchos::parameterList ();
+  params->set("No Nonlocal Changes", true);
+  bool do_params=false;
+
+  if (do_params)
+    sharedNotOwnedMatrix_->fillComplete(params);
+  else
+    sharedNotOwnedMatrix_->fillComplete();
+
+  ownedMatrix_->doExport(*sharedNotOwnedMatrix_, *exporter_, Tpetra::ADD);
+  if (do_params)
+    ownedMatrix_->fillComplete(params);
+  else
+    ownedMatrix_->fillComplete();
+
+  // RHS
+  ownedRhs_->doExport(*sharedNotOwnedRhs_, *exporter_, Tpetra::ADD);
+}
+
+int TpetraSegregatedLinearSystem::solve(stk::mesh::FieldBase * linearSolutionField)
+{
+
+  TpetraLinearSolver *linearSolver = reinterpret_cast<TpetraLinearSolver *>(linearSolver_);
+
+  if ( realm_.debug() ) {
+    checkForNaN(true);
+    if (checkForZeroRow(true, false, true)) {
+      throw std::runtime_error("ERROR checkForZeroRow in solve()");
+    }
+  }
+
+  if (linearSolver->getConfig()->getWriteMatrixFiles()) {
+    writeToFile(eqSysName_.c_str());
+    writeToFile(eqSysName_.c_str(), false);
+  }
+
+  double solve_time = -NaluEnv::self().nalu_time();
+
+  int iters;
+  double finalResidNorm;
+
+  // memory diagnostic
+  if ( realm_.get_activate_memory_diagnostic() ) {
+    NaluEnv::self().naluOutputP0() << "NaluMemory::TpetraSegregatedLinearSystem::solve() PreSolve: " << eqSysName_ << std::endl;
+    realm_.provide_memory_summary();
+  }
+
+  const int status = linearSolver->solve(
+      sln_,
+      iters,
+      finalResidNorm,
+      realm_.isFinalOuterIter_);
+
+  solve_time += NaluEnv::self().nalu_time();
+
+  if (linearSolver->getConfig()->getWriteMatrixFiles()) {
+    writeSolutionToFile(eqSysName_.c_str());
+    ++eqSys_->linsysWriteCounter_;
+  }
+
+  copy_tpetra_to_stk(sln_, linearSolutionField);
+  sync_field(linearSolutionField);
+
+  // computeL2 norm
+  Teuchos::Array<double> mv_norm(1);
+  ownedRhs_->norm2(mv_norm());
+  const double norm2 = mv_norm[0];
+
```

Comment:
Does this need an all-reduce?

Reply:
Yes, `Tpetra::MultiVector` computes the global norm of each vector.
By the way this is already being used in `TpetraLinearSystem` since I switched it from using `Tpetra::Vector` to `Tpetra::MultiVector`.

Reply:
@rcknaus this has been modified in latest patch